File: 30DaysOfCode/day12/inheritance.cpp
#include <iostream>
#include <vector>

using namespace std;


class Person{
	protected:
		string firstName;
		string lastName;
		int id;
	public:
		Person(string firstName, string lastName, int identification) {
			this->firstName = firstName;
			this->lastName = lastName;
			this->id = identification;
		}
		void printPerson(){
			cout << "Name: "<< lastName << ", " << firstName << "\nID: " << id << "\n"; 
		}
	
};

class Student :  public Person {
	private:
		vector<int> testScores;  
	public:
        Student(string firstName, string lastName, int id, vector<int> scores) : Person(firstName, lastName, id) {
            testScores = scores;
        }
    
        char calculate() {
            int scoreAvg = 0;
            
            for (int i = 0; i < testScores.size(); i++) {
                scoreAvg += testScores[i];
            }
            scoreAvg /= testScores.size();
            
            if (scoreAvg >= 90 && scoreAvg <= 100)
                return 'O';
            else if (scoreAvg >= 80 && scoreAvg < 90)
                return 'E';
            else if (scoreAvg >= 70 && scoreAvg < 80)
                return 'A';
            else if (scoreAvg >= 55 && scoreAvg < 70)
                return 'P';                
            else if (scoreAvg >= 40 && scoreAvg < 55)
                return 'D';                
            return 'T';
        }
};

int main() {
	string firstName;
  string lastName;
	int id;
  int numScores;
	cin >> firstName >> lastName >> id >> numScores;
  vector<int> scores;
  for(int i = 0; i < numScores; i++){
	  int tmpScore;
	  cin >> tmpScore;
		scores.push_back(tmpScore);
	}
	Student* s = new Student(firstName, lastName, id, scores);
	s->printPerson();
	cout << "Grade: " << s->calculate() << "\n";
	return 0;
}

File: 30DaysOfCode/day12_inheritance/inheritance.php
<?php

class Person {
    protected $firstName, $lastName, $id;
    
    public function __construct($first_name, $last_name, $identification) {
        $this->firstName = $first_name;
        $this->lastName = $last_name;
        $this->id = $identification;
    }
    
    function printPerson() {
        print("Name: {$this->lastName}, {$this->firstName}\n");
        print("ID: {$this->id}\n");
    }
}

class Student extends Person {
    private $testScores;
  
    public function __construct($first_name, $last_name, $id, $scores) {
        parent::__construct($first_name, $last_name, $id);
        $this->testScores = $scores;
    }

    function calculate() {
        $scoreAvg = 0;
        for ($i = 0; $i < count($this->testScores); $i++) {
            $scoreAvg += $this->testScores[$i];
        }
        $scoreAvg /= count($this->testScores);
        
        if ($scoreAvg >= 90 && $scoreAvg <= 100)
            return 'O';
        else if ($scoreAvg >= 80 && $scoreAvg < 90)
            return 'E';
        else if ($scoreAvg >= 70 && $scoreAvg < 80)
            return 'A';
        else if ($scoreAvg >= 55 && $scoreAvg < 70)
            return 'P';                
        else if ($scoreAvg >= 40 && $scoreAvg < 55)
            return 'D';                
        return 'T';
    }
}

$file = fopen("php://stdin", "r");

$name_id = explode(' ', fgets($file));

$first_name = $name_id[0];
$last_name = $name_id[1];
$id = (int)$name_id[2];

fgets($file);

$scores = array_map(intval, explode(' ', fgets($file)));

$student = new Student($first_name, $last_name, $id, $scores);

$student->printPerson();

print("Grade: {$student->calculate()}");
